pgrx/src/pg_catalog/pg_proc.rs

```diff
@@ -0,0 +1,316 @@
+            _ => panic!("unrecognized `ProArgMode`: `{}`", value),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ProKind {
+    Function,
+    Procedure,
+    Aggregate,
+    Window,
+}
+
+impl From<i8> for ProKind {
+    fn from(value: i8) -> Self {
+        match value as u8 {
+            b'f' => ProKind::Function,
+            b'p' => ProKind::Procedure,
+            b'a' => ProKind::Aggregate,
+            b'w' => ProKind::Window,
+
+            // there's just no ability to move forward if given a value that we don't know about
+            _ => panic!("unrecognized `ProKind`: `{}`", value),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ProVolatile {
+    Immutable,
+    Stable,
+    Volatile,
+}
+
+impl From<i8> for ProVolatile {
+    fn from(value: i8) -> Self {
+        match value as u8 {
+            b'i' => ProVolatile::Immutable,
+            b's' => ProVolatile::Stable,
+            b'v' => ProVolatile::Volatile,
+
+            // there's just no ability to move forward if given a value that we don't know about
+            _ => panic!("unrecognized `ProVolatile`: `{}`", value),
+        }
+    }
+}
+
+pub enum ProParallel {
+    Safe,
+    Restricted,
+    Unsafe,
+}
+
+impl From<i8> for ProParallel {
+    fn from(value: i8) -> Self {
+        match value as u8 {
+            b's' => ProParallel::Safe,
+            b'r' => ProParallel::Restricted,
+            b'u' => ProParallel::Unsafe,
+
+            // there's just no ability to move forward if given a value that we don't know about
+            _ => panic!("unrecognized `ProParallel`: `{}`", value),
+        }
+    }
+}
+
+impl Drop for PgProc {
+    fn drop(&mut self) {
+        // SAFETY: We have a valid pointer and this just decrements the reference count.
+        // This will generally get resolved by the end of the transaction anyways,
+        // but Postgres strongly recommends you do not do that.
+        unsafe { pg_sys::ReleaseSysCache(self.inner.as_ptr()) }
+    }
+}
+
+impl PgProc {
+    /// Construct a new [`PgProc`] from a known function [`pg_sys::Oid`].  If the specified oid is not
+    /// a function, we return [`None`].
+    pub fn new(pg_proc_oid: pg_sys::Oid) -> Option<PgProc> {
+        unsafe {
+            // SAFETY:  SearchSysCache1 will give us a valid HeapTuple or it'll return null.
+            // Either way, using NonNull::new()? will make the right decision for us
+            let entry = pg_sys::SearchSysCache1(
+                pg_sys::SysCacheIdentifier_PROCOID as _,
+                pg_proc_oid.into_datum().unwrap(),
+            );
+            let inner = NonNull::new(entry)?;
+            Some(PgProc { inner })
+        }
+    }
+
+    /// Owner of the function
+    pub fn proowner(&self) -> pg_sys::Oid {
+        // SAFETY:  `proowner` has a NOT NULL constraint
```

Comment:
Are these actually safety constraints? If so, `get_attr` should be an `unsafe fn`. Otherwise, you should remove `SAFETY:` from the comment (here and below)

Reply:
Those notes were about the .unwrap() not panicking.  I'll change the wording.